spring-integration-core/src/main/java/org/springframework/integration/annotation/EndpointId.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.annotation;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * When used alongside an EIP annotation (and no {@code @Bean}) specifies the bean name of
+ * the consumer bean with the handler bean being {@code id.handler} (for a consuming
+ * endpoint) or {@code id.source} for a message source (e.g. inbound channel adapter).
+ * <p>
+ * When there is also a {@code @Bean} annotation, this is the name of the consumer or
+ * source polling bean (the handler or source gets the normal {@code @Bean} name). When
+ * using on a {@code MessageHandler @Bean}, it is recommended to name the bean
+ * {@code foo.handler} when using {@code @EndpointId("foo"}. This will align with
+ * conventions in the framework. Similarly, for a message source, use
+ * {@code @Bean("bar.source"} and {@code @EndpointId("bar")}.
+ *
+ * @author Gary Russell
+ *
+ * @since 5.0.4
+ */
+@Target({ElementType.METHOD, ElementType.ANNOTATION_TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+@Inherited
```

Comment:
Not sure that `@Inherited` is OK here. 
If you override method and then register it in the application context with some Messaging Annotation, you definitely have to have its own distinct endpoint id. Therefore `@EndpointId` is required on the overridden method.

WDYT?

Reply:
Looks like this is wrong on several of the EIP annotations...
```
 * <p>Note that this meta-annotation type has no effect if the annotated
 * type is used to annotate anything other than a class.  Note also
 * that this meta-annotation only causes annotations to be inherited
 * from superclasses; annotations on implemented interfaces have no
 * effect.
```

Reply:
Hm. So, sounds like fixing that won't cause any breaking changes.
Isn't it ?